### ISSUE TITLE
Multiple templates support

### ISSUE THREAD

**Issue by robertovelasco** [mamund/hal-forms#10]
In the current specification there is only a valid template per hal-forms response.
It is interesting to have multiple template support.
The missing piece is how to link each template with its link. 

I can imagine a working example like this one:

``` json
{
  "_links" : {
    "self" : {
      "href" : "http://api.example.org/filter"
    },
    "edit" : {
      "href" : "http://api.example.org/edit"
    }
  },
  "_templates" : {
    "default" : {
      "title" : "Filter",
      ...
    },
    "edit" : {
      "title" : "Edit",
      "method":"put",
      ...
    }
  }
}
```

Where the template 'default' always uses the link with rel 'self' and the rest of templates will use a link with the same name (template 'edit' and link 'edit' for example) or 'self' as fallback.

**Comment by mamund** [comment]
@robertovelasco 

**Single REL, Multiple Templates**
IIRC, on the HAL list we discussed whether the response could include more than one `template` element for *the same URI*. for example, a single URI (`rel="http://example.org/rels/customers"`) might point to human-readable documentation that says:

> http://example.org/customers
> Client apps can use the URL associated with the link-rel-value to do the following:
> - GET the list (or single element, depending on the associated URL)
> - if the associated URL ends in a slash ('/'), POST a new customer record to the list
> - if the associated URL ends in an identifier ('/123/), PUT an updated customer record to the URL
> - if the associated URL ends in an identifier ('/123'), DELETE the existing customer record from the list

In the above case, a single HAL-FORMS response might look like this:
```
{
  "_links" : {
    "self" : {
      "href" : "http://api.example.org/rels/customers"
    }
  },
  "_templates" : {
    "default" : {
      "title" : "Read Customers",
      "method" : "get"
    },
    "create" : {
      "title" : "Add Customer",
      "method":"post",
      ...
    },
    "update" : {
      "title" : "Edit Customer",
      "method":"put",
      ...
    },
    "delete" : {
      "title" : "Remove Customer",
      "method":"delete",
      ...
    }
  }
}
```

**Multiple RELs, Multiple Templates**
It's also possible that a single HAL-FORMS document could return ALL the available forms for a response (or for the app -- e.g. a single call to allow the client to cache all the possible actions).  then, the `_links` section will have multiple entries and there needs to be a way to match a `link` element with a `template` (one or more?) element. 

I'd like to support both scenarios, if possible, without adding too much complexity for parser/render tooling. We can discuss on the list or here.

**Comment by robertovelasco** [comment]
Ok, for the 'Multiple RELs, Multiple Templates' case I can see two possible solutions:
1. curies style template names.
2. include an optional new attribute with link's rel

###curies style template names

```json
{
  "_links" : {
    "self" : {
      "href" : "http://api.example.org/user"
    },
    "orders" : {
      "href" : "http://api.example.org/orders"
    }
  },
  "_templates" : {
    "default" : {
      "title" : "edit",
      ...
    },
    "orders:add" : {
      "title" : "New",
      "method":"post",
      ...
    },
    "orders:filter" : {
      "title" : "Filter",
      "method":"post",
      ...
    }
  }
}
```

Template name contains a reference to a rel: "rel-name:template-name".

###New attribute
Add a new `rel` attribute to the template to indicate the corresponding link.

```json
{
  "_links" : {
    "self" : {
      "href" : "http://api.example.org/user"
    },
    "orders" : {
      "href" : "http://api.example.org/orders"
    }
  },
  "_templates" : {
    "default" : {
      "title" : "edit",
      ...
    },
    "add" : {
      "rel" : "orders",
      "title" : "New",
      "method":"post",
      ...
    },
    "filter" : {
      "rel" : "orders",
      "title" : "Filter",
      "method":"post",
      ...
    }
  }
}
```

**Comment by mamund** [comment]
using curies is an idea -- i'd want to make sure our use of it inside HAL-FORMS conforms to the way they are used within HAL responses.

i will also admit i am a bit worried about "smuggling" new `link` values within HAL-FORMS (ones that do not appear in the associated HAL response). this might not be a big deal, but wanted to put it out there.

i think taking these details to the HAL list will get more eyeballs and surface any problems.

but, yes, we have a couple ways ahead. cool.

**Comment by dschulten** [comment]
@robertovelasco are the links http://api.example.org/user and http://api.example.org/orders in your examples extension link relation types? They look a lot like URLs for the user resource and the related orders list, that is why I am asking. If I understand the hal-form spec correctly, the hal form is supposed to be retrieved at the URL which identifies the extension relation type.

Assuming you have a user resource which points to related orders using an extension relation type
http://api.example.com/rels/orders:

``` javascript
{
  "_links": {
    "http://api.example.com/rels/orders" : { <-- URL where hal-form comes from
      "href" : "http://api.example.org/orders" 
    }
  }
  "name": "Sheldon Cooper"
} 
``´

**Comment by dschulten** [comment]
@mamund in the single rel, multiple templates case: would the four _template keys default, create, update, delete be fixed values with a technical meaning, or can implementations make up template keys?

Also, I don't understand how the client can construct the PUT and DELETE Uris, based on this:

```
{
  "_links" : {
    "self" : {
      "href" : "http://api.example.org/rels/customers"
    }
  },
  "_templates" : {
    "update" : {
      "title" : "Edit Customer",
      "method":"put",
      ...
    },
    "delete" : {
      "title" : "Remove Customer",
      "method":"delete",
      ...
    }
  }
}
```

Suggestion: let there be two different hal-form documents - one for single customer urls (it has PUT and DELETE templates) and another one for the collection (POST). Since the single customer and the customers are two different resources, it does not feel right to describe interactions with a single customer not on its own rel, but on the rel which is actually the collection resource.

**Comment by mamund** [comment]
> I don't understand how the client can construct the PUT and DELETE Uris

that's because clients DO NOT construct URIs in HAL-FORMS. the URL used for these actions is in the HAL document (it might be a templated URI, too). HAL-FORMS are where _additional_ metadata about the link/action is kept such as method, arguments, and human-readable prompts and titles.

and the `"self" : { "href" : "..."}` link at the top of the document is the location of the HAL-FORMS document. this was the value of the `rel` in the HAL document.

hope this clears things up.

**Comment by mamund** [comment]
I have no plans to constrain the keys in the `_templates` object except to include a `"default"` template. This is the only domain-agnostic template in the HAL-FORMS document.

The remaining keys in the `_templates` object are meant to be domain-specific. They'll be explained in the human-readable documentation in the same way HAL works today. And, as HAL works today, programmers will need to translate that human-readable documentation into programming code for the client app.

**Comment by dschulten** [comment]
> and the "self" : { "href" : "..."} link at the top of the document is the location of the HAL-FORMS document. this was the value of the rel in the HAL document.

What confuses me is that all four templates do not seem to work on the same url. Rather there are probably two different resources, the one to post to (/customers), and the one to put and delete against (/customers/42). If they all belonged to the same rel in the preceding hal response, they would share the same resource. So I guess the typical crud case still requires two rels and two hal-forms documents with the single rel/multiple action idiom, right?

**Comment by mamund** [comment]
first, snot sure what example your working from on the whole /customers and customers/42.
second, there is nothing in HAL that promises any CRUD-like assumptions or behaviors. 
third, if the HAL URL (not the HAL-FORMS URL) is *templated* there might be human-readable documentation that offers some advice on how to *resolve* that template which could result in a target URL of /customers/ or /customers/42.

for example:
if the HAL url is "http://api.example.org/customers/{id}"
and the HAL-FORMS document contains two templates lableled "create", "update", and "remove" the runtime client MAY resolve the `{id}` template with a value of the current item (e.g. "id=42") for the "update" and "remove" and resolve the `{id}` with a value of "" for the "create" template.

this is all wild speculation since i don't have the documentation here to read and interpret.

finally, all this vagueness falls away when:
- you only emit a "default" template  in the HAL-FORMS document (my current spec)
- you don't use templated URLs as the target address within the HAL document.

does this help?

**Comment by mamund** [comment]
@robertovelasco 

I've got a question on your *Multiple REL, Multiple Templates* extension -- basically around the suggested process model for HAL-FORMS [1]

In the current spec, a) the client finds a link-rel with a URI within a HAL document, b) the client makes a call to that link-rel URI requesting a HAL-FORMS document, and, if one is returned c) uses the "default" template within that HAL-FORMS document to form a body to send to the server using the indicated HTTP method (from the template) to the URL _in the HAL document_.

With a HAL-FORM that has Multiple REL values, Multiple Templates, I am not sure I understand how the processing model is expected to work. Can you write up a short walk-through here?

Questions that I'm not clear on:
- how does the client initially discover this HAL-FORM?
- how does the client know which link-rel (from the `_links` section) to use when processing the HAL-FORM?
- how does the client know which key in the `_template` to select from the HAL-FORM response?
- how does the client know which HAL document URL to use as the target when sending the formatted request?

Something along lines similar to [1] would be cool.

cheers.

[1] http://mamund.site44.com/misc/hal-forms/#_suggested_process_flow_for_hal_forms_documents

**Comment by dschulten** [comment]
> if the HAL url is "http://api.example.org/customers/{id}"

Yep, that clears things up :-)

I know that http methods are not equivalents to CRUD. However, the template keys create, update and delete kind of put me on a track ;-)

**Comment by dschulten** [comment]
@robertovelasco One way to get all forms for the links of the current resource could be to use the profile link, which according to mike kelly should be used to describe the current resource.

The templates could be organized by the rels in the described hal representation.

Assuming we come from a  /store resource containing a link with rel /rels/customers, a link with /rels/offers and a link with rel profile having a /rels/store href:

Hal forms:
{
... self is /rels/store...
_templates: {
   "http://api.example.org/rels/customers" : {
     default: {... filter customers},
     orders: {... post an order},
  ... more rels pointing to product offers etc.
}

@mamund in a previous design you had a similar model with rels as template key, why did you switch to "default"?

**Comment by mamund** [comment]
@dschulten 

the first pass didn't support multiple templates. when adding possible support for more than one template, it was decided to make sure at least *one* template would be a domain-agnostic "default" to make it easy for ppl who didn't want to support multiple AND didn't want domain-specific knowledge required to use HAL-FORMS.

**Comment by mamund** [comment]
@robertovelasco @dschulten 

IME, supporting the _Multiple REL, multiple Template_ is going to be a very high bar for anything other than bespoke client apps. I don't see a clear algorithm for a client app to navigate a single HAL-FORMS document that contains **all possible actions and templates**. I'd be happy to be wrong on this.

I'd like to see the client-side HAL-FORMS processing model you plan to implement in Spring that will support this _Multiple REL, multiple Template_ scenario. Without seeing (and reviewing/vetting) that process model, I think you're heading into a problem area with no clear plan to stay out of trouble.

**Comment by gillarramendi** [comment]
Regarding _Multiple REL, multiple Template_ we agree all solutions are difficult to understand and implement in server and client side.
For now we prefer to focus efforts on solving other issues, specially what is presented below.

### Suggested process flow for dynamic forms

First of all, it’s important to note that within our requirements, we need that forms/templates are specific to a user and the state of the resource. So it means that we need to generate a form definition instance in real time, and it will change depending on user roles and the state of the resource.

As far as we know the recommended behavior of the hal-forms specification doesn’t work this way. It generates _static_ form definitions that are generic for all users. 

In this situation, there are some non-CRUD cases that are difficult to cover, for example, the refund of a Payment.

```json

GET /api/orders/1
Accept: application/hal+json

{
  "_links": {
    "self": "/api/orders/1",
    
    "example:payments": {
      "href" : "/api/orders/1/payments/{paymentId}"
    },
    
    "curies": [{
      "example": "http://localhost/rel/{rel}"
    }]
  }
}

GET http://localhost/rel/payments
Accept: application/prs.hal-forms+json

{
  "_links": {
    "self": "http://localhost/rel/payments"
  }
  "_templates":{
    "default": {                  "<-- Pay the order. Included only if the order has not been paid yet!",
      "title": "Pay order",
      "method" : "POST",
      "properties" : [
        {"name" : "creditCard", "required" : true, "value" : "", "prompt" : "Credit card Number"}
      ]
    },
    "refund": {                   "<-- Included only if the order has been paid."
      "title": "Refund payment",
      "method": "PUT"
      
    }
  }
}
```
As you can see the in the HAL-FORM approach the possible actions against a specific payment id are always the same (is a static definition) but we would like to change the response depending on the status of the payment. We consider this fits with the HATEOAS philosophy where the server side offers only the possible actions to the clients. 

The same happens with the possible values of a form fields, that can change. I mean, for instance when we have a select to show the users bank accounts, the possible accounts are specific to that user and in consequence the form definition is particular for that user.


In order to implement a dynamic form definition, we consider that `edit-form` and `create-form` rels defined in IANA could cover this requirement:. 

This is the suggested process flow:

1. Servers emit HAL responses that contain links **(prefered with `edit-form` and `create-form` rels but not only)** that return HAL-FORMS documents.
2. Clients parse the HAL response and (either on-demand or in pre-fetch mode) request the HAL-FORMS documents.
3. When a HAL-FORMS document is returned by the server, clients use this information to render an input UI for humans when needed.
4. Clients collect the completed user inputs and, based on the value of contentType, craft a valid request to send to the server and execute that request.

`edit-form` and `create-form` link can contain resource specific ids, for example:

```json
"_links": {
    "self": {
      "href": "http://api.example.org/orders/1",
    },
    "edit-form": {
      "href": "http://api.example.org/orders/1/edit-form",
    },
    "example:refunds": {
      "href": "http://api.example.org/orders/1/payments/refund-form",
    },
   "curies":....
}
```
The response of the `edit-form` link will be a `hal-forms` document with a form created for the task with id=1. It will contain all the possible values and actions the client can invoke, depending on its state.

One pending question is where to include the url of the action. In `hal-forms` it is in the HAL document, but in this approach we have to include it within the `hal-forms` response, as it is proposed within IANA RFC (https://tools.ietf.org/html/rfc6861). 
There are two options:

- Include new `href` property in the template.
- A link in the `_links` section with a known `rel`.

You can see that it is similar to how an HTML form is discovered and invoked.

What do you think about this approach? 

We understand that this question is different to the topic on this issue, @mamund if you prefer we can open another issue.

**Comment by mamund** [comment]
@gillarramendi 

> As far as we know the recommended behavior of the hal-forms specification doesn’t work this way. It generates static form definitions that are generic for all users.

that's not at all correct. the specification makes no constraints on how servers populate or emit HAL-FORMS documents. 

> we would like to change the response depending on the status of the payment

Nothing in the HAL-FORMS spec prevents you from doing this. 

> The same happens with the possible values of a form fields, that can change.

Again, this is totally fine according to the HAL-FORMS spec. nothing new here.

>  One pending question is where to include the url of the action. In hal-forms it is in the HAL document, but in this approach we have to include it within the hal-forms response

I strongly advise against this approach. There is no need to "smuggle" target URLs in a HAL-FORMS document. This kind of implementation runs counter to the way the HAL specification is written, hides target URLs from HAL doc consumers, and needlessly complicates HAL-FORMS processing.

If you want to pursue this design change, start a new issue and we can continue to discussion there.

again, i asked about a *client-side* processing model for _Multiple REL, Multiple Template_ cases. i appreciate you are focused on server-side implementation details for the Spring work and remind you that HAL-FORMS does not make any constraints on how *servers* compose and emit them (as long as the output is valid HAL-FORMS). The much harder work is writing HAL-FORMS _clients_ and that is where I think some real thinking needs to be done before you create these extensions.

cheers.

**Comment by gillarramendi** [comment]
Thanks @mamund for the clarification and sorry for the misunderstanding.

I think we are talking about different subjects in the same thread so I will try to answer the different points below:

- Multiple templates support
That’s the original subject of this issue. Basically we have been talking about two kind of multiple templates:

  - Single REL & Multiple Templates
Regarding your specification (For this release, the only valid value for key is "default") it seems that it is not possible to have more than one template within one HAL-FORM response. We consider it would be interesting to support more than one template within each HAL-FORM response because we have different actions related to one rel.
Do you agree to include it in the spec?


  - Multiple REL & Multiple Templates
We agree that this case is maybe too complex to manage so we consider the _Single REL & Multiple Templates_ is enough at this point.



- URL of the action within HAL-FORM
We have understood the reason behind. This feature is not in the current spec and it does not fit properly also with HAL specification so we are not going  to request this feature any more.

- Process flow for HAL-FORMs
  Sorry, I think we haven’t explained this requirement properly. I will try to explain it below.

  As you said we know that the process flow defined in the HAL-FORM spec is not mandatory.  Actually we are not going to propose any change to the current spec (maybe to the doc), but we think it would be advisable to define or extend the process flow proposal for some kind of forms.

  Taking into account this subject is not related to that thread at all, we have created a new issue just for that: https://github.com/mamund/hal-forms/issues/14

**Comment by mamund** [comment]
@gillarramendi 

thanks for the reply and sorry for my delay in response.

> Regarding your specification (For this release, the only valid value for key is "default") it seems that it is not possible to have more than one template within one HAL-FORM response.

right, the existing spec is a bit too constrained there, I'll be changing it to say "the only _defined_ value for key is "default". Other values MAY exist and clients SHOULD ignore any values they do not understand.

> we consider the Single REL & Multiple Templates is enough at this point.
understood.

i'll address the other thread shortly.

**Comment by mamund** [comment]
all:

with the addition of the Query String Registry features (http://rwcbook.github.io/hal-forms/#_the_hal_forms_query_string_registry) I think most of these elements are resolved -- at least in some workable way even if it is not the ways suggested here.

i'm going to close this issue and bring the repo in sync with the RWCBook spec. we can address shortcomings/bugs once that is done.

thanks.
